gen/list_test.go

```diff
@@ -0,0 +1,220 @@
+				ListOfStrings: []string{},
+			},
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
+				{
+					ID:    1,
+					Value: wire.NewValueList(wire.ValueListFromSlice(wire.TBinary, []wire.Value{})),
+				},
+			}}),
+			[]string{},
+		},
+	}
+	for _, tt := range tests {
+		// required
+		w, err := tt.p.ToWire()
+		assert.NoError(t, err, "failed to serialize: %encodedValue", tt.p)
+		assert.True(t, wire.ValuesAreEqual(tt.offTheWire, w))
+
+		// Round trip them all.
+		freshestV, b := assertBinaryRoundTrip(t, w, tt.desc)
+		assert.True(t, b, "failed round trip")
+		assert.True(t, tt.p.Equals(tt.p))
+		// Allocate a new instance to serialize from Thrift representation.
+		x := new(tc.ListOfRequiredPrimitives)
+		if assert.NoError(t, x.FromWire(freshestV), tt.desc) {
+			assert.Equal(t, tt.expect, x.ListOfStrings)
+		}
+	}
+}
+
+func TestListRequiredFromWire(t *testing.T) {
+	tests := []struct {
+		desc       string
+		offTheWire wire.Value
+		expected   *tc.ListOfRequiredPrimitives
+	}{
+		{
+			"empty list field decodes into an empty slice",
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
+				{
+					ID:    1,
+					Value: wire.NewValueList(wire.ValueListFromSlice(wire.TBinary, []wire.Value{})),
+				},
+			}}),
+			&tc.ListOfRequiredPrimitives{
+				ListOfStrings: []string{},
+			},
+		},
+	}
+	for _, tt := range tests {
+		x := new(tc.ListOfRequiredPrimitives)
+		if assert.NoError(t, x.FromWire(tt.offTheWire), tt.desc) {
+			assert.Equal(t, tt.expected, x)
+			_, b := assertBinaryRoundTrip(t, tt.offTheWire, tt.desc)
+			assert.True(t, b, "failed round trip")
+		}
+	}
+}
+
+// Error if required list is missing in the wire representation.
+func TestListRequiredFromWireError(t *testing.T) {
+	tests := []struct {
+		desc       string
+		offTheWire wire.Value
+		wantError  string
+	}{
+		{
+			"empty list field decodes into empty",
+			wire.NewValueStruct(wire.Struct{}),
+			"field ListOfStrings of ListOfRequiredPrimitives is required",
+		},
+	}
+	for _, tt := range tests {
+		x := new(tc.ListOfRequiredPrimitives)
+		err := x.FromWire(tt.offTheWire)
+		if assert.Error(t, err, tt.desc) {
+			assert.Equal(t, tt.wantError, err.Error())
+		}
+	}
+}
+
+// TestListOptionalToWire tests optional serialization cases.
+func TestListOptionalToWire(t *testing.T) {
+	tests := []struct {
+		desc       string
+		offTheWire wire.Value
+		p          *tc.ListOfOptionalPrimitives
+	}{
+		{
+			"optional nil list: no list field",
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{}}),
+			&tc.ListOfOptionalPrimitives{
+				ListOfStrings: nil,
+			},
+		},
+		{
+			"optional empty list: with list field",
+			wire.NewValueStruct(wire.Struct{Fields: []wire.Field{
+				{
+					ID:    1,
+					Value: wire.NewValueList(wire.ValueListFromSlice(wire.TBinary, []wire.Value{})),
+				},
+			}}),
+			&tc.ListOfOptionalPrimitives{
+				ListOfStrings: []string{},
+			},
+		},
+	}
+	for _, tt := range tests {
+		assertRoundTrip(t, tt.p, tt.offTheWire, tt.desc)
```

Comment:
This doesn't need to be a round-trip test. It can be,

    t.Run(tt.desc, func(t *testing.T) {
        var got tc.ListOfOptionalPrimitives
        require.NoError(t, got.FromWire(tt.give), "failed to decode")
        assert.Equal(t, tt.want, got, "values did not match")
    })

The contract we're testing here stops at the wire.Value. It doesn't hurt to
round-trip it, but it's not necessary.

Reply:
yup, i will leave a comment how full round-trip does more than what we need.